### PR TITLE
machineconfig: Set RPS mask for both containers and host

### PR DESCRIPTION
**Pull request by marcel-apf**
Add a pre-start OCI hook setting the CPUs RPS mask to the performance profile's reserved CPUs.

To-do:
- Some tests

Signed-off-by: Marcel Apfelbaum <marcel@redhat.com>

### CODE REVIEW

File: build/assets/configs/99-high-performance-hooks.json
Comment (cynepco3hahue) [inline build/assets/configs/99-high-performance-hooks.json:5]: probably all occurrences of high-performance should be renamed to be low-latency, I remember @MarSik asked for the same thing for me

File: build/assets/scripts/high-performance-hooks.sh
Comment (cynepco3hahue) [inline build/assets/scripts/high-performance-hooks.sh:9]: `[[ -n "${mask}"  ]] &&  [[ -n "${ns}" ]]` two spaces after &&

**Review by cynepco3hahue — COMMENTED** [review]
Except some small comments looks good for now

File: build/assets/configs/99-high-performance-hooks.json
Comment (marcel-apf) [reply build/assets/configs/99-high-performance-hooks.json:5]: @MarSik please ack, I don;t care one way or another :)

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:9]: thanks, changed

File: build/assets/scripts/high-performance-hooks.sh
Comment (fromanirh) [inline build/assets/scripts/high-performance-hooks.sh:9]: If this fails there is no point in continuing. OTOH if we do we should "only" get a bunch of errors somewhere

File: build/assets/scripts/high-performance-hooks.sh
Comment (fromanirh) [inline build/assets/scripts/high-performance-hooks.sh:7]: is it worth to exit early if mask exists and it is equal to "0" (= rps disabled)?

File: build/assets/scripts/high-performance-hooks.sh
Comment (fromanirh) [inline build/assets/scripts/high-performance-hooks.sh:11]: I wonder if it makes sense to split this script in two, one which identifies the NS and executes the real script in the right namespace, and another one which just does the work. Something like

```
$ cat assets/scripts/run-in-ns.sh
#!/usr/bin/env bash

pid=$(jq '.pid' /dev/stdin)
mask=$1
what=$2
ns=$(ip netns identify "${pid}")

[[ -n "${pid}" ]] && [[ -n "${mask}" ]] && [[ -n "${ns}" ]] || exit 0

ip netns exec "${ns}" ${what}
```
and
```
$ cat assets/scripts/high-performace-hooks.sh
#!/usr/bin/env bash
mount -o remount,rw /sys
find /sys/devices -type f -name rps_cpus -exec sh -c "echo ${mask} | cat > {}" \;
mount -o remount,ro /sys
```
Whatever's easier to write and maintain, I guess.

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:9]: yep, I better bail out [no yet though about tightening things up]

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:7]: in the future we may want to disable rps... and the default value is ) anyway, I would leave it as is

Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh:7]: works for me!

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:9]: I will add a "set -e" to the script to bail out if anything goes wrong, would be good enough?

Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh:9]: using `set -e` is a good idea in general, but it seems our case it's a bit more complex.
let's consider the last three commands, which is the real deal here.
1. if remounting sysfs rw fails, we do want to bail out (but keep doing will do no harm besides wasted work)
2. if find fails, we still want to remount sysfs ro!
3. if remounting sysfs ro fails, we have no way to recover anyway so nothing to do here

Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:9]: right, I'll be more precise

File: build/assets/scripts/high-performance-hooks.sh
Comment (fromanirh) [inline build/assets/scripts/high-performance-hooks.sh:9]: nice!

**Comment by marcel-apf** [comment]
Addressed @fromanirh;s comments
Didn't rename yet the files to low-latency-..., waiting for @MarSik ack
Add support for dynamic host network device rps updates

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:9]: :)

File: build/assets/scripts/high-performance-hooks.sh
Comment (fromanirh) [inline build/assets/scripts/high-performance-hooks.sh:10]: any reason to use a different find here with respect to https://github.com/openshift-kni/performance-addon-operators/pull/450/commits/bde1fc3aa43f0b62e1c72259d47ab9c6d24b663f#diff-b550247b819f015b31bacae7c601e2ef1045d94b6ff48ecb29c3cb362578be75R306 ?

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (fromanirh) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:47]: this seems to be used just once in line 165 below. If so why can't we just call this "99-netdev-rps.rules"  ?

**Review by fromanirh — COMMENTED** [review]
very nice improvements, kudos!

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:10]: funny thing, find doesn't find the rps_cpus files  in /sys/class/net/ dir.
It does find them in  /sys/class/net/vethxyz/ or in /sys/devices/

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:47]: I have this weird rule to use the same coding standard in the file, especially the micro patterns, so readers would not wonder why is different. I agree there is no other need for that.
I am OK one way or the other.

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:10]: is something about an implementation difference of the sys/class vs /sys/devices

Comment (cynepco3hahue) [inline build/assets/scripts/high-performance-hooks.sh:10]: After the @fromanirh comment regarding error handling, IMHO we should add -e and add a trap that will call to `ip netns exec "${ns}" mount -o remount,ro /sys`, because if the command
`ip netns exec "${ns}" find /sys/devices -type f -name rps_cpus -exec sh -c "echo ${mask} | cat > {}" \;` will fail but the remount command after it will succeed, the hook will exit with 0 code, and from the cri-o perspective it will mean that hook succeeded, so a user can get unexpected results

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (cynepco3hahue) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:305]: should this service start before the kubelet service?

Comment (cynepco3hahue) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:305]: nit: use upper case letters for acronyms - getRPSUnitOptions

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (cynepco3hahue) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:245]: nit: use upper case letters for acronyms - getCRIOHooksConfigContent

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:10]: This is on purpose, I am interested to return as less errors as possible, if the hook fails I want the pod to continue to run.  I prefer the pod to be running with the rps mask not set, than not running at all

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:305]: The server is only triggered by the udev rule

Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:305]: will change, thanks

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:245]: Missed that one, thanks

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (fromanirh) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:47]: I'm ok following the existing conventions, so we can keep it this way.

File: build/assets/scripts/high-performance-hooks.sh
Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh:10]: ok, no problem. Worth a comment too. Others (or our future selves) may well wonder the same thing in the future.

Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh:10]: Will add a comment in the script, thanks

**Comment by marcel-apf** [comment]
Fixes:
 - Rename files to low-latency... (@cynepco3hahue )
 - Correct naming for functions (@cynepco3hahue )
 - Add a needed comment in script (@fromanirh )

**Comment by marcel-apf** [comment]
Fixed a minor typo in a script comment

File: build/assets/configs/99-high-performance-hooks.json
Comment (marcel-apf) [reply build/assets/configs/99-high-performance-hooks.json]: changed, thanks

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: (Is not connected to the kubelet at all)

Comment (cynepco3hahue) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: I thought it should configure all masks before starting all containers?

File: build/assets/scripts/high-performance-hooks.sh
Comment (cynepco3hahue) [reply build/assets/scripts/high-performance-hooks.sh]: hm if you expect some functionality and you do not get it silently it can lead to unexpected behavior, but I will not block on it

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: well, this is a new approach ending in the same way.
The udev will trigger an event for each new net device, so is enough to handle only the device.

The previous approach failed as it didn't take into account other net devices created after the container started.


This approach has 2 steps
1. Set the rps mask for each pod when the pod is started
2. Set the rps mask for each host net device as soon as is created

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: yeah, is a hack. we will push for a sane RPS interface, sadly is the best it can be done at current time.

Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh]: The question is: even if we properly exit with error and we log the reason, where do the log ends and how we can check it? If it ends up somewhere we can check (crio logs?) it's worth improving. Even something trivial like well defined exit codes for the failure conditions can be enough.

Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh]: In general, what happens if a hook exits with error?

Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: The point is we don't actually want to error out. We can use the Logger to send the logs to the syslog, but is not necessary since is the caller's responsibility to catch the stderr of the script and log it.

The current approach does the best it can:
1. if it can't remount the sysfs as rw, exits with 0 so it will let the pod continue to run, it makes no sense to punish some random pod because of this.
2. At this point we go for the rps_cpus files write, even if it doesn't succeed, we still need to try to unmount and the pod should still continue to run.
3. The last try to remount as ro is tricky, we shouldn't let an unprivileged pod to run with sysfs rw. But if so, since is the last script line, it will return a non-zero error code and the pod will not run.

All of the above is (should be) logged by the caller.

Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh]: > The point is we don't actually want to error out. We can use the Logger to send the logs to the syslog, but is not necessary since is the caller's responsibility to catch the stderr of the script and log it.

True. Do we know where do these logs end? crio logs?

> The current approach does the best it can:
> 
>     1. if it can't remount the sysfs as rw, exits with 0 so it will let the pod continue to run, it makes no sense to punish some random pod because of this.

Well, this is debatable, and that's indeed why we are discussing. Guaranteed pods which opt in to the behaviour we are implementing here expect some preconditions. If we fail to set up the RPS mask these preconditions aren't guaranteed anymore. At this point keep going with the pod start up may or may be not the right thing to do but what @cynepco3hahue AFAIU raised (and if so I totally agree) is that we should never *silently* fail to comply with the pod expectations.

So if the hook cannot properly set up the environment the guarantees aren't met anymore and this fact should be evident somewhere. Which is indeed responsability of the caller (crio) to log, but we need to know where this happens and how we can check later on.

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (MarSik) [inline pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:315]: Make sure the unit is executed in a deterministic order. Eg. use After and Before as needed (After=network, Before=kubelet? <-- not sure about this one).

File: build/assets/scripts/low-latency-hooks.sh
Comment (MarSik) [inline build/assets/scripts/low-latency-hooks.sh:1]: This is executed for all the pods, not just guaranteed ones, right?

File: build/assets/scripts/high-performance-hooks.sh
Comment (MarSik) [reply build/assets/scripts/high-performance-hooks.sh]: But this hook is called for all the pods in the system, right? I see two issues:

1) Last time we used cri-o code approach instead of wrapper, because the OCP maintainers were worried about scalability (what happens when huge amount of pods starts at the same time). This we can't probably fix atm.
2) A failing hook prevents the start of pods, we must be super careful to not compromise the whole node

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (MarSik) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: Oh, isn't the service also started once on boot?

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: > In general, what happens if a hook exits with error?

The container ends up in status: CreateContainerError
And in the pod status we see the exact error:
  Warning  Failed          1s               kubelet, ocp47-worker-1.demo.lab  Error: container create failed: time="2020-11-26T08:26:48Z" level=error msg="container_linux.go:366: starting container process caused: process_linux.go:472: container init caused: Running hook #0:: error running hook: exit status 1, stdout: , stderr: /usr/local/bin/high-performance-hooks.sh: line 13: /bin/d: Read-only file system\n"

Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: > True. Do we know where do these logs end? crio logs?
> 

The pod event status has the detailed output, please see the above exampe

Comment (fromanirh) [reply build/assets/scripts/high-performance-hooks.sh]: Thanks, container ending up in `CreateContainerError` matches my expectations and it's indeed better than silently fail to meet the expectations.

Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: > Well, this is debatable, and that's indeed why we are discussing. Guaranteed pods which opt in to the behaviour we are implementing here expect some preconditions. If we fail to set up the RPS mask these preconditions aren't guaranteed anymore. At this point keep going with the pod start up may or may be not the right thing to do but what @cynepco3hahue AFAIU raised (and if so I totally agree) is that we should never _silently_ fail to comply with the pod expectations.
> 

We can opt to return error in this case, that means is possible no container would start and the node will appear as "Not Ready". But if the consensus is that is a critical matter I will change it to return error.


> So if the hook cannot properly set up the environment the guarantees aren't met anymore and this fact should be evident somewhere. Which is indeed responsability of the caller (crio) to log, but we need to know where this happens and how we can check later on.

If the script is returning with exit 0, no interim error is logged and sensing something to syslog does not work either.

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go]: No, the service is not started on boot, is started only by udev on net dev creation.

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: > But this hook is called for all the pods in the system, right? I see two issues:
> 
> 1. Last time we used cri-o code approach instead of wrapper, because the OCP maintainers were worried about scalability (what happens when huge amount of pods starts at the same time). This we can't probably fix atm.

Nope, we can't, it will be a short delay on each pod startup. The "price" to pay for the low-latency.
But, as you pointed out, the systems requiring RT are quite static, running a few pods.

> 2. A failing hook prevents the start of pods, we must be super careful to not compromise the whole node
Exactly what I said! 

We have 2 points of view and we need a decision:
If the pod hook fails to do its job it will affect the Guaranteed Pods latency.
So what should we do?
 - Deny it to run can and will compromise the node, for example it may not allow an operator pod to run. 
 - Allow it to run anyway will compromise the Guaranteed Pods latency.

File: build/assets/scripts/low-latency-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/low-latency-hooks.sh:1]: Right

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (marcel-apf) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:315]: It does not run on system boot, the dependencies are not relevant.

File: build/assets/scripts/high-performance-hooks.sh
Comment (marcel-apf) [reply build/assets/scripts/high-performance-hooks.sh]: > Thanks, container ending up in `CreateContainerError` matches my expectations and it's indeed better than silently fail to meet the expectations.

Right, we don't have a security risk.

Our Achilles' heel remains if the writing the rps_cpus fails and the remount to ro succeeded. In this case the hook returns with 0. I found no way to log something if the hook succeeded

File: pkg/controller/performanceprofile/components/machineconfig/machineconfig.go
Comment (MarSik) [reply pkg/controller/performanceprofile/components/machineconfig/machineconfig.go:315]: Ack, I did not know about systemd unit templates feature. A nifty trick!

File: build/assets/scripts/low-latency-hooks.sh
Comment (MarSik) [reply build/assets/scripts/low-latency-hooks.sh:1]: We need a scale test.. starting 250 pods at the same time and such.

**Comment by marcel-apf** [comment]
Renamed "crioHools" -> ociHooks

**Comment by marcel-apf** [comment]
Improved the low-latency-hooks.sh script:
 - check the input
 - add logs for every step
 - handle the  case when the container has /sys rw already (fx bug that returned it to rw)

File: build/assets/scripts/low-latency-hooks.sh
Comment (fromanirh) [inline build/assets/scripts/low-latency-hooks.sh:27]: this looks *very* nice and well thought, kudos.

**Comment by cynepco3hahue** [comment]
@fromanirh @MarSik Can you please review again and unhold once the PR looks good to you.

**Review by fromanirh — APPROVED** [review]
/lgtm
/hold cancel

**Pull request merged**
